### PR TITLE
[multistage][hotfix] fix type cast handling and value literal gen

### PR DESCRIPTION
**Pull request by walterddr**

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java [hunk 1]
```diff
@@ -448,21 +449,30 @@ private static void rewriteRangeOperator(Expression range, FilterKind kind, int
...
+  /** @return field data type extracted from the expression. null if we can't determine the type. */
+  private static @Nullable FieldSpec.DataType getDataType(Expression expression, Schema schema) {
+    if (expression.getType() == ExpressionType.IDENTIFIER) {
+      String column = expression.getIdentifier().getName();
+      FieldSpec fieldSpec = schema.getFieldSpecFor(column);
+      if (fieldSpec != null && fieldSpec.isSingleValueField()) {
+        return fieldSpec.getDataType();
+      }
+    } else if (expression.getType() == ExpressionType.FUNCTION
+        && "cast".equalsIgnoreCase(expression.getFunctionCall().getOperator())) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java:461]: You might also want to handle alias (`as`)?

File: pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java [hunk 2]
```diff
@@ -448,21 +449,30 @@ private static void rewriteRangeOperator(Expression range, FilterKind kind, int
...
+      String targetTypeLiteral = expression.getFunctionCall().getOperands().get(1).getLiteral().getStringValue()
+          .toUpperCase();
+      FieldSpec.DataType dataType;
+      if ("INTEGER".equals(targetTypeLiteral)) {
+        dataType = FieldSpec.DataType.INT;
+      } else if ("VARCHAR".equals(targetTypeLiteral)) {
+        dataType = FieldSpec.DataType.STRING;
+      } else {
+        dataType = FieldSpec.DataType.valueOf(targetTypeLiteral);
+      }
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java:472]: Consider making a util for this part, and make `DataTypeConversionFunctions.cast()` and `CastTransformFunction` use the util. Currently they already have different behavior.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorUtils.java [hunk 3]
```diff
@@ -51,6 +52,7 @@ private OperatorUtils() {
    * @return Canonicalize form of the input function name
    */
   public static String canonicalizeFunctionName(String functionName) {
+    functionName = StringUtils.replace(functionName, " ", "_");
```
Comment (Jackie-Jiang) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorUtils.java:55]: We should just remove the space. See `TransformFunctionFactory.canonicalize()` for example

File: pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java (hunk 2 above)
Comment (walterddr) [reply pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java:472]: they are not equivalent. conversion function operates `@ScalarFunction` which runs on `PinotDataType` (physical data type) and here it runs on `FieldSpec.DataType`.

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java [hunk 4]
```diff
@@ -283,14 +290,22 @@ private Object[][] provideTestSql() {
...
 
         // Test DISTINCT
         //   - distinct value done via GROUP BY with empty expr aggregation list.
         new Object[]{"SELECT a.col2, a.col3 FROM a JOIN b ON a.col1 = b.col1 "
             + " WHERE b.col3 > 0 GROUP BY a.col2, a.col3"},
+
+        // Test optimized constant literal.
+        new Object[]{"SELECT col1 FROM a WHERE col3 > 0 AND col3 < -5"},
+        new Object[]{"SELECT COALESCE(SUM(col3), 0) FROM a WHERE col1 = 'foo' AND col1 = 'bar'"},
+        new Object[]{"SELECT SUM(CAST(col3 AS INTEGER)) FROM a HAVING MIN(col3) BETWEEN 1 AND 0"},
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:307]: For the query at line 307, can we add a test with `CAST to LONG` ?

File: pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java [hunk 5]
```diff
@@ -93,18 +94,21 @@ public Expression optimize(Expression expression, @Nullable Schema schema) {
...
+          FieldSpec.DataType dataType = getDataType(lhs, schema);
+          if (dataType != null && dataType.isNumeric()) {
+            switch (kind) {
+              case EQUALS:
+              case NOT_EQUALS:
+                return rewriteEqualsExpression(expression, kind, dataType, rhs);
+              case GREATER_THAN:
+              case GREATER_THAN_OR_EQUAL:
+              case LESS_THAN:
+              case LESS_THAN_OR_EQUAL:
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java:107]: BETWEEN ?

File: pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java [hunk 6]
```diff
@@ -448,21 +449,31 @@ private static void rewriteRangeOperator(Expression range, FilterKind kind, int
...
-      // Expression can not be a column name.
-      return false;
+  /** @return field data type extracted from the expression. null if we can't determine the type. */
+  @Nullable
+  private static FieldSpec.DataType getDataType(Expression expression, Schema schema) {
+    if (expression.getType() == ExpressionType.IDENTIFIER) {
+      String column = expression.getIdentifier().getName();
+      FieldSpec fieldSpec = schema.getFieldSpecFor(column);
+      if (fieldSpec != null && fieldSpec.isSingleValueField()) {
+        return fieldSpec.getDataType();
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java:459]: if `fieldSpec` is null, then probably it implies that column  / identifier does not exist in the schema and we should throw error ?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorUtils.java [hunk 7]
```diff
@@ -51,6 +52,7 @@ private OperatorUtils() {
    * @return Canonicalize form of the input function name
    */
   public static String canonicalizeFunctionName(String functionName) {
+    functionName = StringUtils.remove(functionName, " ");
```
Comment (siddharthteotia) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorUtils.java:55]: Ideally this should not even be allowed imo. Can you give an example ?

**Review by siddharthteotia — APPROVED** [review]

File: pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java (hunk 4 above)
Comment (walterddr) [reply pinot-query-runtime/src/test/java/org/apache/pinot/query/runtime/QueryRunnerTest.java:307]: CAST TO LONG is not supported. LONG is not a SQL standard  type -- it has to be CAST TO BIGINT

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorUtils.java (hunk 7 above)
Comment (walterddr) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/OperatorUtils.java:55]: calcite automatically converts multi-part functions like `IS NOT NULL` to that in SqlOp table.

**Comment by walterddr** [comment]
will follow up with the comments in a separate PR especially refactoring out the CAST type conversion into a utility class.

**Pull request merged**